bluepyentity/register.py

```diff
@@ -0,0 +1,472 @@
+        refined = self._format_attributes(self._definition)
+        refined = self._wrap_fields_with_ids(refined)
+
+        # Fields to not store in Nexus
+        refined.pop("brainRegion", None)
+
+        return self._with_defaults(refined)
+
+    @property
+    def possible_ids(self):
+        """Return the fields that possibly contains nexus ids"""
+        return self._schema_definition["ids"]
+
+    @property
+    def resource(self):
+        """Access the resource."""
+        return self._resource
+
+    @property
+    def schema(self):
+        """Schema for the entity."""
+        # TODO: figure out how to get this from Nexus and use it to validate types etc.
+        return self._schema_definition["schema"]
+
+    @property
+    def type(self):
+        """Return the Nexus type of the resource."""
+        return self._definition["type"]
+
+    @cached_property
+    def _schema_definition(self):
+        return get_entity_definition(self.type)
+
+    def register(self):
+        """Register the resource in Nexus."""
+        if existing := self._find_existing():
+            bucket = self._forge._store.bucket  # pylint: disable=protected-access
+            raise RuntimeError(
+                f"Similar '{self.type}' definition already exists in project "
+                f"'{bucket}' with an id of: '{existing.id}'"
+            )
+
+        try:
+            schema_id = self._forge._model.schema_id(self.type)  # pylint: disable=protected-access
+        except ValueError:
+            schema_id = None
+
+        try:
+            self._forge.register(self.resource, schema_id=schema_id)
+        except Exception as err:
+            raise RuntimeError(err) from err
+
+    def _check_if_valid(self):
+        self._check_required()
+        # NOTE: should this be removed, if schemas are handled by the backend?
+        # Should extra attributes be allowed? Currently, the backend allows them.
+        # self._check_schema()
+
+    def _check_required(self):
+        """Check that the required items are defined."""
+        missing = self.required - set(self.definition)
+        assert not missing, f"Missing attributes: {missing}"
+
+    def _check_schema(self):
+        disallowed = set(self.definition) - set(self.schema)
+        disallowed -= {"type"}
+        assert not disallowed, f"Disallowed attributes: {disallowed}"
+
+    def _create_resource(self):
+        """Create a kgforge Resource of the definition."""
+        definition = self.definition
+        definition["distribution"] = self._get_distribution()
+
+        return kgforge.core.Resource.from_json(definition)
+
+    def _format_attributes(self, definition):
+        """Format needed values in the definition."""
+        return definition
+
+    def _find_existing(self):
+        """Find resource matching the definition in Nexus."""
+        found = self._forge.search({"type": self.type}, cross_bucket=True)
```

Comment:
Minor: This, by default, uses the sparql endpoint, which is expensive. Is it possible to use the elastic endpoint instead?
https://github.com/BlueBrain/nexus-forge/blob/e4edecac49dad1cc0255f0f936ac06cf95c36648/kgforge/specializations/stores/bluebrain_nexus.py#L687-L689

Reply:
This was a way to circumvent the issue that some fields, such as the `circuitConfigPath` for circuits was not searchable. 

To be honest, I don't like this, nor the idea of helper code having to decide which one should be differentiating field that makes each instance unique. It should be the job of the back end and it should be defined in specs/schemas etc. 

The more I think of it, there's currently one unique field defined, `id`, which we should check, nothing more. If the users register duplicates, then I would assume it is a matter of them not finding the instances in the Knowledge Graph. If they aren't able to find stuff there, then we have bigger issues than us registering duplicates. 

Rant over, summa summarum, I will remove `_find_existing`, and `_is_equal_to_nexus_resource`.